File: Algo_Ds_Notes/Cycle_Sort/Cycle_Sort.java

import java.io.*;
import java.util.*;

public class CycleSort
{
    static void sort (int a[], int n)
    {
        int writes = 0, start, element, pos, temp, i;
        
        // Find position where we put the element.
        for (start = 0; start <= n - 2; start++)
        {
            element = a[start];
            pos = start;
           
           /*Find position where we put the element. We basically 
           count all smaller elements on right side of element. */
           for (i = start + 1; i < n; i++)
           if (a[i] < element)
                pos++;
            
            // If element is already in correct position     
            if (pos == start)
                continue;
            
            // ignore all duplicate elements      
            while (element == a[pos])
                pos += 1;
            
            // put the element to it's right position
            if (pos != start) {
                temp = element;
                element = a[pos];
                a[pos] = temp;
                writes++;
            }
           
            // Rotate rest of the cycle  
            while (pos != start) 
            {
                pos = start;
                for (i = start + 1; i < n; i++)
                  if (a[i] < element)
                    pos += 1;
                
                // ignore all duplicate elements 
                while (element == a[pos])
                    pos += 1;
                
                // put the item to it's right position 
                if (element != a[pos]) {
                    temp = element;
                    element = a[pos];
                    a[pos] = temp;
                    writes++;
                }
            }
        }
    }
    
    public static void main (String[]args)
    {
       int n;
        Scanner s = new Scanner(System.in);
        System.out.print("Enter no. of elements you want in array:");
        n = s.nextInt();
        int a[] = new int[n];
        System.out.println("Enter all the elements:");
        for(int i = 0; i < n; i++)
            a[i] = s.nextInt();
        sort (a, n);
        System.out.println ("After sort, array : ");
        for (int j = 0; j < n; j++)
            System.out.println (a[j]);
    }
}

/*Output:
Enter no. of elements you want in array:5                                                                                 
Enter all the elements:                                                                                                   
12                                                                                                                        
45                                                                                                                        
-7                                                                                                                        
0                                                                                                                         
10                                                                                                                        
After sort, array :                                                                                                       
-7                                                                                                                        
0                                                                                                                         
10                                                                                                                        
12                                                                                                                        
45 */

File: Algo_Ds_Notes/Cycle_Sort/Cycle_Sort.cs
/// C# program for cycle sort 
using System; 
  
class CycleSort { 
      
    // Function to implement Cycle sort 
    public static void cycleSort(int[] arr, int n) 
    {
        int writes = 0; 
        // traverse through array elements and find the right position  
        for (int cycle_start = 0; cycle_start <= n - 2; cycle_start++) 
        {  
            int item = arr[cycle_start]; 
            int pos = cycle_start; 
            for (int i = cycle_start + 1; i < n; i++) 
                if (arr[i] < item) 
                    pos++; 
                    
            //If item is in correct position
            if (pos == cycle_start) 
                continue; 
            
            //Ignoring all duplicate elements
            while (item == arr[pos]) 
                pos += 1; 
            
            //Keeping the item in its correct position
            if (pos != cycle_start) { 
                int temp = item; 
                item = arr[pos]; 
                arr[pos] = temp; 
                writes++; 
        } 
  
            while (pos != cycle_start) { 
                pos = cycle_start; 
              
                // Find position where we put the element 
                for (int i = cycle_start + 1; i < n; i++) 
                    if (arr[i] < item) 
                        pos += 1; 
              
               //Ignoring duplicate elements
                while (item == arr[pos]) 
                    pos += 1;
              
                //Keeping the item in its right position
                if (item != arr[pos]) { 
                    int temp = item; 
                    item = arr[pos]; 
                    arr[pos] = temp; 
                    writes++; 
                } 
            } 
        } 
    } 
  
public static void Main() 
    { 
        int size;
        size= Convert.ToInt32(Console.ReadLine());  
        int[] array = new int[size];
        for (int i=0; i<size; i++) 
    	      array[i] = Convert.ToInt32(Console.ReadLine()); 
        int n = arrray.Length;
        CycleSort(array, n); 
        Console.Write("After sorting  : ")
        for (int i = 0; i < n; i++) 
            Console.Write(array[i] + " "); 
    } 
} 
//SAMPLE INPUT AND OUTPUT
// 7 
// 4 7 9 5 3 9 4
// After sorting : 3 4 4 5 7 9 9 
// 5
// 5 1 7 3 2
// After sorting : 1 2 3 5 7
